### PR TITLE
The knife bootstrap raises `ERROR: TypeError: no implicit conversion of nil into String`  when Chef::Config[:validation_key] is nil.

### PR DESCRIPTION
**Pull request by sawanoboly**
It will break compatibility of my plugin.

Please check that it is not nil before file existence check.

### CODE REVIEW

**Comment by btm** [comment]
At first it looked like you duplicated the entire previous context for the test, but only changed one line in the before block. Then I noticed you flipped two of the three expects.

@sawanoboly what is your plugin? Are you specifically setting validation key to nil, because it has a default value (except in the case of using chef-zero).

/cc @lamont-granquist because he wrote this chunk of code recently.

File: lib/chef/knife/bootstrap.rb [hunk 1]
```diff
@@ -313,7 +313,7 @@ def run
 
         # chef-vault integration must use the new client-side hawtness, otherwise to use the
         # new client-side hawtness, just delete your validation key.
-        if chef_vault_handler.doing_chef_vault? || !File.exist?(File.expand_path(Chef::Config[:validation_key]))
+        if chef_vault_handler.doing_chef_vault? || Chef::Config[:validation_key] && !File.exist?(File.expand_path(Chef::Config[:validation_key]))
```
Comment (lamont-granquist) [inline lib/chef/knife/bootstrap.rb]: i like parens around the `&&` expression written like this so i don't have to think about if operator precedence matters.

**Comment by lamont-granquist** [comment]
LGTM.  Somehow I think I convinced myself it would fail gracefully for the nil case, but clearly it does not.

File: spec/unit/knife/bootstrap_spec.rb [hunk 2]
```diff
@@ -593,6 +593,20 @@ def certificates
...
+    context "when the validation_key is nil" do
+      before do
+        # this tests runs the old code path where we have a validation key, so we need to pass that check for some plugins
+        Chef::Config[:validation_key] = nil
+      end
+
+      it "creates the client (and possibly adds chef-vault items)" do
+        expect(knife_ssh).to receive(:run)
+        expect(knife.client_builder).not_to receive(:run)
+        expect(knife.chef_vault_handler).not_to receive(:run).with(node_name: knife.config[:chef_node_name])
```
Comment (jdmundrawala) [inline spec/unit/knife/bootstrap_spec.rb]: Is receiving `:run` ok if node_name is something other than `knife.config[:chef_node_name]`

File: spec/unit/knife/bootstrap_spec.rb [hunk 3]
```diff
@@ -593,6 +593,20 @@ def certificates
...
       end
     end
+
+    context "when the validation_key is nil" do
+      before do
+        # this tests runs the old code path where we have a validation key, so we need to pass that check for some plugins
+        Chef::Config[:validation_key] = nil
+      end
+
+      it "creates the client (and possibly adds chef-vault items)" do
```
Comment (btm) [inline spec/unit/knife/bootstrap_spec.rb]: This test does something different than the test in the previous context, so the name needs to be different. From what I gather, this is "creates the client and does not run client_builder or the chef_vault_handler"

File: spec/unit/knife/bootstrap_spec.rb (hunk 2 above)
Comment (lamont-granquist) [inline spec/unit/knife/bootstrap_spec.rb]: No, it shouldn't receive either of those with any args.

Comment (lamont-granquist) [inline spec/unit/knife/bootstrap_spec.rb]: Looks like copypasta of a mistake of mine earlier in the file.

**Comment by jdmundrawala** [comment]
https://github.com/chef/chef/pull/2999

**Comment by sawanoboly** [comment]
@btm 
My plugin is [knife-zero](https://github.com/higanworks/knife-zero)

@jdmundrawala Thanks for cleanup!

**Pull request closed without merging**